File: 2015/05.py
# --- Day 5: Doesn't He Have Intern-Elves For This? ---

vowels = "aeiou"
bads   = ["ab", "cd", "pq", "xy"]
part1 = part2 = 0

with open("input/05.txt") as f:

  for line in f:
    # part 1
    v_cnt = 0
    double = bad = False
    for v in vowels:
      v_cnt += line.count(v)
    for i in range (0, len(line) - 1):
      if line[i] == line[i+1]:
        double = True
    for b in bads:
      if line.find(b) > -1:
        bad = True
    if v_cnt > 2 and double and not bad: 
      part1 += 1

    # part 2
    pairs = repeats = False
    for i in range (0,len(line)-2):
      a,b,c = line[i:i+3]
      if a == c:
        repeats = True
      if line.count(a+b) > 1:
        pairs = True
    if repeats and pairs:
      part2 += 1

  print part1
  print part2

File: 2015/05.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

func isNice(word string) int {
	vowels := 0
	for _, v := range "aeiou" {
		vowels += strings.Count(word, string(v))
	}
	if vowels < 3 {
		return 0
	}

	doubles := 0
	for i := 0; i < len(word)-1; i++ {
		if word[i] == word[i+1] {
			doubles++
		}
	}
	if doubles == 0 {
		return 0
	}

	for _, v := range []string{"ab", "cd", "pq", "xy"} {
		if strings.Contains(word, v) {
			return 0
		}
	}
	return 1
}

func isReallyNice(word string) int {
	repeats, duplicates := 0, 0

	for i := 0; i < len(word)-2; i++ {
		if word[i] == word[i+2] {
			repeats++
		}
		if strings.Index(word[i+2:len(word)], word[i:i+2]) > -1 {
			duplicates++
		}
	}
	if repeats > 0 && duplicates > 0 {
		return 1
	}
	return 0
}

func main() {
	file, err := os.Open("input/05.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	nice, reallyNice := 0, 0

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		text := scanner.Text()
		nice += isNice(text)
		reallyNice += isReallyNice(text)
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println(nice)
	fmt.Println(reallyNice)
}
